### PR TITLE
[TS Migration] Migrate CheckboxWithLabel.js component to TypeScript

### PR DESCRIPTION
**Pull request by JKobrynski**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25048
PROPOSAL: N/A

### Tests

1. Log into the app
2. Navigate to the Settings page
3. Navigate to the Wallet page
4. Click the 'Add bank account button'
5. Select 'Debit card'
6. You should see the `CheckboxWithLabel` component on the bottom (_I accept the Expensify Terms of Service_)
7. Make sure it looks and works properly

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as **Tests** section above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android-compressed.webm](https://github.com/Expensify/App/assets/44364829/238670fb-d555-422c-95e9-db8294a7d889)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[chrome-compressed.webm](https://github.com/Expensify/App/assets/44364829/3fcdefde-4bfe-4f94-a609-8bdde62e816d)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/44364829/dd50a5e1-94d1-4ba6-833e-420c6ae11d00

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/44364829/99be233d-22be-42d3-b74b-3ca3bb36fec5

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/44364829/35f40eed-e575-4b2c-a338-3527a89cd35c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/44364829/c2484842-cb6c-43e1-89df-9b8edf4ae7da

</details>

### CODE REVIEW

File: src/components/CheckboxWithLabel.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    /** The default value for the checkbox */
+    defaultValue?: boolean;
+
+    /** The ID used to uniquely identify the input in a Form */
+    /* eslint-disable-next-line react/no-unused-prop-types */
+    inputID?: string;
+
+    /** Saves a draft of the input value when used in a form */
+    // eslint-disable-next-line react/no-unused-prop-types
+    shouldSaveDraft?: boolean;
```
Comment (rezkiy37) [inline src/components/CheckboxWithLabel.tsx:55]: Why do we need this one?

File: src/components/Checkbox.tsx [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
 import useStyleUtils from '@styles/useStyleUtils';
 import useThemeStyles from '@styles/useThemeStyles';
 import CONST from '@src/CONST';
-import ChildrenProps from '@src/types/utils/ChildrenProps';
 import Icon from './Icon';
 import * as Expensicons from './Icon/Expensicons';
 import PressableWithFeedback from './Pressable/PressableWithFeedback';
 
-type CheckboxProps = ChildrenProps & {
+type CheckboxProps = {
```
Comment (fabioh8010) [inline src/components/Checkbox.tsx:11]:
```suggestion
type CheckboxProps = Partial<ChildrenProps> & {
```

What about this?

File: src/components/CheckboxWithLabel.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      }
+    | {
+          /** Component to display for label */
+          LabelComponent: ComponentType;
+
+          /** Text that appears next to check box
+           * If LabelComponent is provided, label is not required
+           */
+          label?: string;
+      };
```
[the remark this answers is not in the document]
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx:28]: This won't work 😄 TypeScript won't let me use these props saying they don't exist

File: src/components/CheckboxWithLabel.tsx (hunk 1 above)
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx:55]: I am not sure, but it was there before I started migrating it and it had that eslint disable line 😅 do you think we should remove it?

Comment (fabioh8010) [reply src/components/CheckboxWithLabel.tsx:55]: It's necessary when used inside `Form` component, same as `inputID`.

File: src/components/CheckboxWithLabel.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        onInputChange(newState);
+        setIsChecked(newState);
+    };
+
+    return (
+        <View style={style}>
+            <View style={[styles.flexRow, styles.alignItemsCenter, styles.breakWord]}>
+                <Checkbox
+                    isChecked={isChecked}
+                    onPress={toggleCheckbox}
```
Comment (fabioh8010) [inline src/components/CheckboxWithLabel.tsx:81]:
```suggestion
                    onPress={toggleCheckbox}
                    label={props.label}
```

Why did you remove `label`?

File: src/components/CheckboxWithLabel.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kubabutkiewicz) [reply src/components/CheckboxWithLabel.tsx:28]: I think I don't understand 😅 
Correct me if I am wrong but if we will pass `label` prop to this component we should not pass `LabelComponent` right? 
If yes you can do something like [in Button component](https://github.com/Expensify/App/blob/main/src/components/Button/index.tsx)

[the remark this answers is not in the document]
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx:28]: If I define it like you suggested, it won't let me destructure these two props, the error says: 

`Property 'LabelComponent' does not exist on type 'CheckboxWithLabelProps'.`
`Property 'label' does not exist on type 'CheckboxWithLabelProps'.`

File: src/components/CheckboxWithLabel.tsx (hunk 4 above)
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx:81]: It's not a valid prop for `Checkbox`

File: src/components/CheckboxWithLabel.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    {errorText = '', isChecked: isCheckedProp = false, defaultValue = false, onInputChange = () => {}, LabelComponent, label, accessibilityLabel, style, value}: CheckboxWithLabelProps,
+    ref: ForwardedRef<View>,
+) {
+    const styles = useThemeStyles();
+    // We need to pick the first value that is strictly a boolean
+    // https://github.com/Expensify/App/issues/16885#issuecomment-1520846065
+    const [isChecked, setIsChecked] = useState(() => [value, defaultValue, isCheckedProp].find((item) => typeof item === 'boolean'));
+
+    const toggleCheckbox = () => {
+        const newState = !isChecked;
```
Comment (pac-guerreiro) [inline src/components/CheckboxWithLabel.tsx:71]: Just a nitpick, but I don't see a benefit of defining a const for something so easy to determine which only add one extra line of code

File: src/components/CheckboxWithLabel.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    };
+
+    return (
+        <View style={style}>
+            <View style={[styles.flexRow, styles.alignItemsCenter, styles.breakWord]}>
+                <Checkbox
+                    isChecked={isChecked}
+                    onPress={toggleCheckbox}
+                    style={[styles.checkboxWithLabelCheckboxStyle]}
+                    hasError={Boolean(errorText)}
```
Comment (pac-guerreiro) [inline src/components/CheckboxWithLabel.tsx:83]:
```suggestion
                    hasError={!!errorText}
```

File: src/components/CheckboxWithLabel.tsx (hunk 5 above)
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx]: Right, done!

File: src/components/CheckboxWithLabel.tsx (hunk 6 above)
Comment (JKobrynski) [reply src/components/CheckboxWithLabel.tsx]: Updated!

**Comment by JKobrynski** [comment]
@pac-guerreiro feel free to take another look 😄

**Comment by pac-guerreiro** [comment]
@JKobrynski LGTM 👍

**Review by blazejkustra — APPROVED** [review]

**Review by aimane-chnaif — COMMENTED** [review]
NAB:
https://github.com/Expensify/App/pull/32814/files#diff-00add3b2d3d53f4116f40f4684fabbaf701bb3fa7a35a7a462571d231d09d335R94

```diff
-           role={CONST.ACCESSIBILITY_ROLE.CHECKBOX}
+           role={CONST.ROLE.CHECKBOX}
```

**Review by aimane-chnaif — APPROVED** [review]

**Comment by JKobrynski** [comment]
@mountiny no engineer here either, can you help? 😄

**Review by arosiclair — APPROVED** [review]
